libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,379 @@
+                    hpx::util::allocator_deleter<other_allocator>{alloc});
+
+                new (p.get()) shared_state{std::forward<S_>(s), a};
+                st = p.release();
+
+                // Eagerly start the work received until this point.
+                //
+                // P1897r3 says "When start is called on os2 [the operation
+                // state resulting from connecting the ensure_started_sender to
+                // a receiver], call execution::start(os [the operation state
+                // resulting from connecting S to ensure_started_receiver])",
+                // which would indicate that start should be called later.
+                // However, to fulfill the promise that ensure_started actually
+                // eagerly submits the sender S we call start already here.
+                st->start();
+            }
+
+            ensure_started_sender(ensure_started_sender const&) = default;
+            ensure_started_sender& operator=(
+                ensure_started_sender const&) = default;
+            ensure_started_sender(ensure_started_sender&&) = default;
+            ensure_started_sender& operator=(ensure_started_sender&&) = default;
+
+            template <typename R>
+            struct operation_state
+            {
+                std::decay_t<R> r;
+                hpx::intrusive_ptr<shared_state> st;
+
+                template <typename R_>
+                operation_state(R_&& r, hpx::intrusive_ptr<shared_state> st)
+                  : r(std::forward<R_>(r))
+                  , st(std::move(st))
+                {
+                }
+
+                void start() noexcept
+                {
+                    st->add_continuation(std::move(r));
+                }
+            };
+
+            template <typename R>
+            operation_state<R> connect(R&& r) &&
+            {
+                return {std::forward<R>(r), std::move(st)};
+            }
+
+            template <typename R>
+            operation_state<R> connect(R&& r) &
+            {
+                return {std::forward<R>(r), st};
+            }
+        };
+    }    // namespace detail
+
+    HPX_INLINE_CONSTEXPR_VARIABLE struct ensure_started_t final
+      : hpx::functional::tag_fallback<ensure_started_t>
+    {
+    private:
+        template <typename S,
+            typename Allocator = hpx::util::internal_allocator<>>
+        friend constexpr HPX_FORCEINLINE auto tag_fallback_invoke(
+            ensure_started_t, S&& s, Allocator const& a = {})
+        {
+            return detail::ensure_started_sender<S, Allocator>{
+                std::forward<S>(s), a};
+        }
+
+        template <typename S, typename Allocator,
+            typename NewAllocator = hpx::util::internal_allocator<>>
+        friend constexpr HPX_FORCEINLINE auto tag_fallback_invoke(
+            ensure_started_t, detail::ensure_started_sender<S, Allocator> s,
+            Allocator const& = {})
```

Comment:
@hkaiser continuing from https://github.com/STEllAR-GROUP/hpx/pull/5274/files#r626815356 here.

Sorry that was unclear. There's a mistake here, and this is what I meant to have here:
```suggestion
            NewAllocator const& = {})
```
With that change, my previous explanation might make more sense. `NewAllocator` is the type of an allocator the the new `ensure_started_sender` would use, if it weren't for the fact that this overload simply returns the predecessor sender unchanged.

Reply:
I changed this to what I originally intended. Does it make sense to you to simply ignore the new allocator?

Reply:
Hmmm. Wouldn't it be better to return a new sender that relies on the new allocator (if the allocator is different, that is)? What would be the point of this API otherwise? Please note that I'm not sure I have the full picture here, so I'm likely wrong...

Reply:
That's what I'm not entirely sure about. It depends entirely on why the user specifies a different allocator here. I guess we can agree that if the allocator is the same there's no reason to create a new sender? That's the main purpose of this overload. I get a sender from somewhere that may or may not be started. I call `ensure_started` to make sure it's started, but if it's already been started there's no need for it to be started again (especially as this algorithm requires heap allocation).

It's the case when the allocators are different that things get interesting. I'd say no allocation is "better" than any custom allocator that they user might have in the sense that it's free, but then I might not see all the use cases of providing a custom allocator, other than for performance reasons.

Reply:
Yes, if the allocator is the same and the provided sender was already started, then this function can simply return its argument. If its the same allocator but the argument was not started, simply start it and return the argument.

If the allocator is different I think you should create a new sender that uses the new allocator. Would going from the CPU realm to the GPU realm be a valid use case for this (I'm not sure)?

Reply:
Yeah, I think that's fair enough. I also don't know if there's a valid use case for it. I'd expect most of our initial use cases to be using the default allocator anyway and we know the behaviour there. For the the cases where the allocators are different we can come back and optimize it if it turns out that's necessary, but for now just creating a new sender sounds like the right thing to do.

Note: when I say "if the predecessor sender has been started" it really just means "if the predecessor sender is an ensure_started_sender". There may be other senders that eagerly submit their work which this algorithm doesn't care about. This algorithm ensures that the sender is connected and the resulting operation state is started (which is independent from whether or not the actual work was submitted).